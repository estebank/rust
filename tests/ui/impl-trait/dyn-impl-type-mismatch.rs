trait Trait {}
struct Struct;
impl Trait for Struct {}
fn foo() -> impl Trait {
    Struct
}
fn main() {
    let a: Box<dyn Trait> = if true {
        Box::new(Struct)
    } else {
        foo() //~ ERROR E0308
    };
    let a: dyn Trait = if true { //~ ERROR E0277
        Struct //~ ERROR E0308
    } else {
        foo() //~ ERROR E0308
    };
}
