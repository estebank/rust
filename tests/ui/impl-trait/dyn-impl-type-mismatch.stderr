error[E0277]: the size for values of type `dyn Trait` cannot be known at compilation time
  --> $DIR/dyn-impl-type-mismatch.rs:13:12
   |
LL |     let a: dyn Trait = if true {
   |            ^^^^^^^^^ doesn't have a size known at compile-time
   |
   = help: the trait `Sized` is not implemented for `dyn Trait`
   = note: all local variables must have a statically known size
   = help: unsized locals are gated as an unstable feature
help: consider borrowing here
   |
LL |     let a: &dyn Trait = if true {
   |            +

error[E0308]: mismatched types
  --> $DIR/dyn-impl-type-mismatch.rs:11:9
   |
LL | fn foo() -> impl Trait {
   |             ---------- the found opaque type
...
LL |         foo()
   |         ^^^^^ expected `Box<dyn Trait>`, found opaque type
   |
   = note:   expected struct `Box<dyn Trait>`
           found opaque type `impl Trait`
   = note: for more on the distinction between the stack and the heap, read https://doc.rust-lang.org/book/ch15-01-box.html, https://doc.rust-lang.org/rust-by-example/std/box.html, and https://doc.rust-lang.org/std/boxed/index.html
help: store this in the heap by calling `Box::new`
   |
LL |         Box::new(foo())
   |         +++++++++     +

error[E0308]: mismatched types
  --> $DIR/dyn-impl-type-mismatch.rs:14:9
   |
LL |         Struct
   |         ^^^^^^ expected `dyn Trait`, found `Struct`
   |
   = note: expected trait object `dyn Trait`
                    found struct `Struct`
   = help: `Struct` implements `Trait` so you could box the found value and coerce it to the trait object `Box<dyn Trait>`, you will have to change the expected type as well

error[E0308]: mismatched types
  --> $DIR/dyn-impl-type-mismatch.rs:16:9
   |
LL | fn foo() -> impl Trait {
   |             ---------- the found opaque type
...
LL |         foo()
   |         ^^^^^ expected `dyn Trait`, found opaque type
   |
   = note: expected trait object `dyn Trait`
               found opaque type `impl Trait`
   = help: you can box the `impl Trait` to coerce it to `Box<dyn Trait>`, but you'll have to change the expected type as well

error: aborting due to 4 previous errors

Some errors have detailed explanations: E0277, E0308.
For more information about an error, try `rustc --explain E0277`.
